src/queue.rs

```diff
@@ -0,0 +1,810 @@
+        mem.write_obj(u32::from(desc_index), used_elem).unwrap();
+        mem.write_obj(len as u32, used_elem.unchecked_add(4))
+            .unwrap();
+
+        self.next_used += Wrapping(1);
+
+        // This fence ensures all descriptor writes are visible before the index update is.
+        fence(Ordering::Release);
+
+        mem.write_obj(self.next_used.0 as u16, used_ring.unchecked_add(2))
+            .unwrap();
+    }
+
+    /// Goes back one position in the available descriptor chain offered by the driver.
+    /// Rust does not support bidirectional iterators. This is the only way to revert the effect
+    /// of an iterator increment on the queue.
+    pub fn go_to_previous_position(&mut self) {
+        self.next_avail -= Wrapping(1);
+    }
+}
+
+#[cfg(test)]
+pub(crate) mod tests {
+    extern crate vm_memory;
+
+    use std::marker::PhantomData;
+    use std::mem;
+
+    pub use super::*;
+    use vm_memory::{GuestAddress, GuestMemoryMmap};
+
+    // Represents a location in GuestMemoryMmap which holds a given type.
+    pub struct SomeplaceInMemory<'a, T> {
+        pub location: GuestAddress,
+        mem: &'a GuestMemoryMmap,
+        phantom: PhantomData<*const T>,
+    }
+
+    // The ByteValued trait is required to use mem.read_obj and write_obj.
+    impl<'a, T> SomeplaceInMemory<'a, T>
+    where
+        T: vm_memory::ByteValued,
+    {
+        fn new(location: GuestAddress, mem: &'a GuestMemoryMmap) -> Self {
+            SomeplaceInMemory {
+                location,
+                mem,
+                phantom: PhantomData,
+            }
+        }
+
+        // Reads from the actual memory location.
+        pub fn get(&self) -> T {
+            self.mem.read_obj(self.location).unwrap()
+        }
+
+        // Writes to the actual memory location.
+        pub fn set(&self, val: T) {
+            self.mem.write_obj(val, self.location).unwrap()
+        }
+
+        // This function returns a place in memory which holds a value of type U, and starts
+        // offset bytes after the current location.
+        fn map_offset<U>(&self, offset: GuestUsize) -> SomeplaceInMemory<'a, U> {
+            SomeplaceInMemory {
+                location: self.location.checked_add(offset).unwrap(),
+                mem: self.mem,
+                phantom: PhantomData,
+            }
+        }
+
+        // This function returns a place in memory which holds a value of type U, and starts
+        // immediately after the end of self (which is location + sizeof(T)).
+        fn next_place<U>(&self) -> SomeplaceInMemory<'a, U> {
+            self.map_offset::<U>(mem::size_of::<T>() as u64)
+        }
+
+        fn end(&self) -> GuestAddress {
+            self.location
+                .checked_add(mem::size_of::<T>() as u64)
+                .unwrap()
+        }
+    }
+
+    // Represents a virtio descriptor in guest memory.
+    pub struct VirtqDesc<'a> {
```

Comment:
I think this would be simpler if you just use `VolatileRef<Descriptor>`. Perhaps with that change SomePlaceInMemory is not needed at all?